File: ML/train_intent_classifier.py
#!/usr/bin/env python3
"""
Intent Classifier Core ML Model Training Script
Generates IntentClassifier.mlmodel from training data
"""

import json
import os
import sys

def check_coremltools():
    """Check if coremltools is installed"""
    try:
        import coremltools as ct
        return ct
    except ImportError:
        print("ERROR: coremltools is not installed")
        print("Install with: pip install coremltools")
        sys.exit(1)

def load_training_data(filepath):
    """Load training data from JSON file"""
    with open(filepath, 'r', encoding='utf-8') as f:
        return json.load(f)

def prepare_training_samples(data):
    """Prepare training samples in format for Create ML"""
    samples = []
    labels = []

    for item in data['training_data']:
        intent_type = item['intent_type']
        for sample in item['samples']:
            samples.append(sample['text'])
            labels.append(intent_type)

    return samples, labels

def train_model(ct, samples, labels, output_path):
    """Train text classifier model using Create ML"""
    print(f"Training with {len(samples)} samples...")

    # Create ML text classifier
    from coremltools.models.text_classifier import create_text_classifier

    # Create training data DataFrame
    import pandas as pd
    df = pd.DataFrame({
        'text': samples,
        'label': labels
    })

    # Train model
    model = create_text_classifier(
        training_data=df,
        text_column='text',
        label_column='label',
        maximum_iterations=50
    )

    # Set model metadata
    model.short_description = "Intent classifier for telecom mobile app (version 2)"
    model.author = "ioscrmapp"
    model.version = "2.0"

    # Save model
    model.save(output_path)
    print(f"Model saved to: {output_path}")

    return model

def validate_model(model, test_inputs):
    """Validate model with test inputs"""
    print("\n=== Model Validation ===")
    for text in test_inputs:
        result = model.predict({'text': text})
        label = result.get('label', 'unknown')
        print(f"Input: '{text}'")
        print(f"  → Predicted: {label}")
        print()

def main():
    # Paths
    script_dir = os.path.dirname(os.path.abspath(__file__))
    training_data_path = os.path.join(script_dir, 'training_data', 'intent_training_data_v2.json')
    output_path = os.path.join(script_dir, 'IntentClassifier_v2.mlmodel')

    # Check coremltools
    ct = check_coremltools()

    # Load training data
    print(f"Loading training data from: {training_data_path}")
    data = load_training_data(training_data_path)
    print(f"Intent types: {data['intent_types']}")

    # Prepare samples
    samples, labels = prepare_training_samples(data)
    print(f"Total training samples: {len(samples)}")

    # Train model
    model = train_model(ct, samples, labels, output_path)

    # Validate with user's problematic input
    test_inputs = [
        "check my data package is enough or not",
        "how much data do i have left",
        "remaining data",
        "my data usage",
        "is my data package enough",
        "check my balance",
        "view offers",
        "travel to hongkong"
    ]
    validate_model(model, test_inputs)

    print("\n=== Training Complete ===")
    print(f"Output model: {output_path}")
    print("\nNext steps:")
    print("1. Replace ioscrmapp/ML/IntentClassifier.mlmodel with the new model")
    print("2. Update UserIntentType enum to add data_usage_query")
    print("3. Update IntentRoutingService to handle data_usage_query")

if __name__ == '__main__':
    main()

File: ioscrmapp/ML/TrainIntentClassifier.swift
#!/usr/bin/env swift

import CreateML
import Foundation

// MARK: - Configuration

let scriptDir = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
let csvURL = scriptDir.appendingPathComponent("training_data/training_data.csv")
let outputURL = scriptDir.appendingPathComponent("IntentClassifier_v2.mlmodel")

// MARK: - Load training data

print("Loading training data from: \(csvURL.path)")
guard FileManager.default.fileExists(atPath: csvURL.path) else {
    fputs("ERROR: Training CSV not found at \(csvURL.path)\n", stderr)
    exit(1)
}

guard let dataTable = try? MLDataTable(contentsOf: csvURL) else {
    fputs("ERROR: Failed to load CSV data\n", stderr)
    exit(1)
}

print("Total rows: \(dataTable.size)")

// MARK: - Split into training and validation sets

let (trainingData, validationData) = dataTable.randomSplit(by: 0.85, seed: 42)
print("Training: \(trainingData.size), Validation: \(validationData.size)")

// MARK: - Train text classifier

print("\nTraining MLTextClassifier (maxEnt algorithm)...")
let classifier = try MLTextClassifier(
    trainingData: trainingData,
    textColumn: "text",
    labelColumn: "label"
)

// MARK: - Evaluation

let trainingAccuracy = (1.0 - classifier.trainingMetrics.classificationError) * 100
let validationAccuracy = (1.0 - classifier.validationMetrics.classificationError) * 100
print(String(format: "Training accuracy: %.2f%%", trainingAccuracy))
print(String(format: "Validation accuracy: %.2f%%", validationAccuracy))

// List precision/recall per class
print("\nPer-class metrics (validation set):")
let evalMetrics = classifier.validationMetrics
print("Classification error: \(evalMetrics.classificationError)")

// MARK: - Write model

print("\nWriting model to: \(outputURL.path)")
try classifier.write(to: outputURL)

// Verify output
let attributes = try FileManager.default.attributesOfItem(atPath: outputURL.path)
let fileSize = attributes[.size] as? Int64 ?? 0
print(String(format: "Model saved: %.2f KB", Double(fileSize) / 1024.0))

// MARK: - Quick test

print("\n=== Quick Validation ===")
let testInputs = [
    "check my data package is enough or not",
    "how much data do i have left",
    "remaining data",
    "my data usage",
    "check my balance",
    "view offers",
    "travel to hongkong",
    "recharge my account"
]

for text in testInputs {
    let prediction = try classifier.prediction(from: text)
    print("'\(text)' -> \(prediction)")
}

print("\n=== Training Complete ===")
